tensorboard/schema.py

```diff
@@ -0,0 +1,262 @@
+
+
+BIG_TENSORS_TABLE = TableSchema(
+    name='BigTensors',
+    columns=[ColumnSchema('rowid', Int64ColumnType()),
+             ColumnSchema('customer_number', Int64ColumnType()),
+             ColumnSchema('tag_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('step_count', Int64ColumnType(), not_null=True),
+             ColumnSchema('tensor', BytesColumnType())],
+    keys=['rowid', 'customer_number', 'tag_id', 'step_count'])
+
+EVENT_LOGS_TABLE = TableSchema(
+    name='EventLogs',
+    columns=[ColumnSchema('rowid', Int64ColumnType()),
+             ColumnSchema('customer_number', Int64ColumnType()),
+             ColumnSchema('run_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('event_log_id', Int64ColumnType()),
+             ColumnSchema('path', StringColumnType(
+                 length=1023), not_null=True),
+             ColumnSchema('offset', Int64ColumnType(), not_null=True)],
+    keys=['rowid', 'customer_number', 'run_id', 'event_log_id'])
+
+
+EXPERIMENTS_TABLE = TableSchema(
+    name='Experiments',
+    columns=[ColumnSchema('customer_number', Int64ColumnType()),
+             ColumnSchema('experiment_id', Int64ColumnType()),
+             ColumnSchema('name', StringColumnType(500), not_null=True),
+             ColumnSchema('description', StringColumnType(
+                 65535), not_null=True),],
+    keys=['rowid', 'customer_number', 'experiment_id'])
+
+PLUGINS_TABLE = TableSchema(
+    name='Plugins',
+    columns=[ColumnSchema('plugin_id', Int64ColumnType()),
+             ColumnSchema('name', StringColumnType(length=255), not_null=True)],
+    keys=['plugin_id'])
+
+RUNS_TABLE = TableSchema(
+    name='Runs',
+    columns=[ColumnSchema('rowid', Int64ColumnType()),
+             ColumnSchema('customer_number', Int64ColumnType()),
+             ColumnSchema('experiment_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('run_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('name', StringColumnType(length=1900),
+                          not_null=True)],
+    keys=['rowid', 'customer_number', 'experiment_id', 'run_id'])
+
+TAGS_TABLE = TableSchema(
+    name='Tags',
+    columns=[ColumnSchema('rowid', Int64ColumnType()),
+             ColumnSchema('customer_number', Int64ColumnType()),
+             ColumnSchema('run_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('tag_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('plugin_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('name', StringColumnType(500)),
+             ColumnSchema('display_name', StringColumnType(500)),
+             ColumnSchema('summary_description', StringColumnType(65535)),],
+    keys=['rowid', 'customer_number', 'run_id', 'tag_id'])
+
+TENSORS_TABLE = TableSchema(
+    name='Tensors',
+    columns=[ColumnSchema('rowid', Int64ColumnType()),
+             ColumnSchema('customer_number', Int64ColumnType()),
```

Comment:
I'm not sure if we need a `customer_number` column here. It isn't part of of the `rowid` bits. If it were, then it might slow down the performance of tensorboard, since requests for the charts created by a single user in a single experiment wouldn't distribute across nodes. My working guess has been that locality is only necessary within a single chart.

Reply:
I think the customer_number is needed if we eventually want to support multi-tenancy. The customer_number prevents row collisions (ensures uniqueness for each customer) and also makes it easy to do IAM. For example, on GCP customer_number can just be ProjectNumber. So now I can easily check that a request has credentials with the proper IAM role for the correct project when reading a specific row.

My assumption has been that SQLite is implicitly single tenant. So customer_number would always be 0.

Databases supporting multiple users should include customer_number in the primary key and data localization.

In the current Spanner schema the key for the tensors table is 'rowid', 'customer_number'. So all tensors for a given run and customer number should be stored sequentially but runs from different customers would be interspersed.

Although for spanner it might make more sense to group all rows for a given customer since you are more likely to read all experiments, runs for a given user than multiple users.